Negative WaitGroup counter in SetDetailedStatus

**Describe the bug**
This happens during `e2e-test`
```
panic: sync: negative WaitGroup counter

goroutine 435 [running]:
sync.(*WaitGroup).Add(0x0, 0x0)
        /usr/local/go/src/sync/waitgroup.go:74 +0x105
sync.(*WaitGroup).Done(...)
        /usr/local/go/src/sync/waitgroup.go:99
github.com/skycoin/skywire/pkg/app/appserver.(*Proc).SetDetailedStatus(0xc000239d40, {0xc0018a21c8, 0x7})
        /skywire/pkg/app/appserver/proc.go:337 +0xc5
github.com/skycoin/skywire/pkg/app/appserver.(*RPCIngressGateway).SetDetailedStatus(0xc00006a540, 0xc000345720, 0x1)
        /skywire/pkg/app/appserver/rpc_ingress_gateway.go:82 +0x119
reflect.Value.call({0xc000399500, 0xc0002165d0, 0x13}, {0xb4fb83, 0x4}, {0xc000857ef8, 0x3, 0x3})
        /usr/local/go/src/reflect/value.go:543 +0x814
reflect.Value.Call({0xc000399500, 0xc0002165d0, 0x0}, {0xc00069fef8, 0x3, 0x3})
        /usr/local/go/src/reflect/value.go:339 +0xc5
net/rpc.(*service).call(0xc0004b5000, 0x0, 0x0, 0xc0002138b0, 0xc0001aa580, 0x0, {0xa5ab80, 0xc000345720, 0x0}, {0xa5afc0, ...}, ...)
        /usr/local/go/src/net/rpc/server.go:377 +0x239
created by net/rpc.(*Server).ServeCodec
        /usr/local/go/src/net/rpc/server.go:474 +0x405
```

--- a/Makefile
+++ b/Makefile
@@ -237,6 +237,24 @@ host-apps-static: ## Build app
 	${STATIC_OPTS} go build -trimpath --ldflags '-linkmode external -extldflags "-static" -buildid=' -o ./apps/ ./cmd/apps/vpn-server
 	${STATIC_OPTS} go build -trimpath --ldflags '-linkmode external -extldflags "-static" -buildid=' -o ./apps/ ./cmd/apps/vpn-client
 
+host-apps-deploy: ## Build app
+	test -d apps && rm -r apps || true
+	mkdir -p ./apps
+	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o ./apps/ ./cmd/apps/skychat
+	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o ./apps/ ./cmd/apps/skysocks
+	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o ./apps/ ./cmd/apps/skysocks-client
+	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o ./apps/ ./cmd/apps/vpn-server
+	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o ./apps/ ./cmd/apps/vpn-client
+
+host-apps-race: ## Build app
+	test -d apps && rm -r apps || true
+	mkdir -p ./apps
+	CGO_ENABLED=1${OPTS} go build ${BUILD_OPTS} -race -o ./apps/ ./cmd/apps/skychat
+	CGO_ENABLED=1${OPTS} go build ${BUILD_OPTS} -race -o ./apps/ ./cmd/apps/skysocks
+	CGO_ENABLED=1${OPTS} go build ${BUILD_OPTS} -race -o ./apps/ ./cmd/apps/skysocks-client
+	CGO_ENABLED=1${OPTS} go build ${BUILD_OPTS} -race -o ./apps/ ./cmd/apps/vpn-server
+	CGO_ENABLED=1${OPTS} go build ${BUILD_OPTS} -race -o ./apps/ ./cmd/apps/vpn-client
+
 # Bin
 bin: ## Build `skywire-visor`, `skywire-cli`
 	${OPTS} go build ${BUILD_OPTS} -o ./ ./cmd/skywire-visor
@@ -270,12 +288,13 @@ bin-static: ## Build `skywire-visor`, `skywire-cli`
 	${STATIC_OPTS} go build -trimpath --ldflags '-linkmode external -extldflags "-static" -buildid=' -o ./skywire-cli  ./cmd/skywire-cli
 	${STATIC_OPTS} go build -trimpath --ldflags '-linkmode external -extldflags "-static" -buildid=' -o ./setup-node ./cmd/setup-node
 
-build-deploy: ## Build for deployment Docker images
+build-deploy: host-apps-deploy ## Build for deployment Docker images
 	${OPTS} go build -tags netgo ${BUILD_OPTS_DEPLOY} -o /release/skywire-visor ./cmd/skywire-visor
 	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o /release/skywire-cli ./cmd/skywire-cli
-	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o /release/apps/skychat ./cmd/apps/skychat
-	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o /release/apps/skysocks ./cmd/apps/skysocks
-	${OPTS} go build ${BUILD_OPTS_DEPLOY} -o /release/apps/skysocks-client ./cmd/apps/skysocks-client
+
+build-race: host-apps-race ## Build for testing Docker images
+	CGO_ENABLED=1 ${OPTS} go build -tags netgo ${BUILD_OPTS} -race -o /release/skywire-visor ./cmd/skywire-visor
+	CGO_ENABLED=1 ${OPTS} go build ${BUILD_OPTS} -race -o /release/skywire-cli ./cmd/skywire-cli
 
 github-prepare-release:
 	$(eval GITHUB_TAG=$(shell git describe --abbrev=0 --tags | cut -c 2-6))

--- a/cmd/skywire-cli/commands/visor/app.go
+++ b/cmd/skywire-cli/commands/visor/app.go
@@ -168,7 +168,7 @@ var setAppSecureCmd = &cobra.Command{
 
 var setAppPasscodeCmd = &cobra.Command{
 	Use:   "passcode <name> <passcode>",
-	Short: "Set app passcode.\n\"remove\" is a special arg to remove the passcode",
+	Short: "Set app passcode.\n               \"remove\" is a special arg to remove the passcode",
 	Args:  cobra.MinimumNArgs(2),
 	Run: func(cmd *cobra.Command, args []string) {
 		passcode := args[1]
@@ -182,7 +182,7 @@ var setAppPasscodeCmd = &cobra.Command{
 
 var setAppNetworkInterfaceCmd = &cobra.Command{
 	Use:   "netifc <name> <interface>",
-	Short: "Set app network interface.\n\"remove\" is a special arg to remove the netifc",
+	Short: "Set app network interface.\n               \"remove\" is a special arg to remove the netifc",
 	Args:  cobra.MinimumNArgs(2),
 	Run: func(cmd *cobra.Command, args []string) {
 		netifc := args[1]
@@ -196,7 +196,7 @@ var setAppNetworkInterfaceCmd = &cobra.Command{
 
 var appLogsSinceCmd = &cobra.Command{
 	Use:   "log <name> <timestamp>",
-	Short: "Logs from app since RFC3339Nano-formatted timestamp.\n\"beginning\" is a special timestamp to fetch all the logs",
+	Short: "Logs from app since RFC3339Nano-formatted timestamp.\n               \"beginning\" is a special timestamp to fetch all the logs",
 	Args:  cobra.MinimumNArgs(2),
 	Run: func(cmd *cobra.Command, args []string) {
 		var t time.Time

--- a/docker/images/visor/Dockerfile
+++ b/docker/images/visor/Dockerfile
@@ -14,7 +14,6 @@ WORKDIR /skywire
 
 RUN apk add --no-cache make git && \
     sh -c /skywire/docker/images/visor/install-preq.sh && \
-    make host-apps && \
     make build-deploy && \
     mkdir -p /apps && \
     cp ./apps/vpn-server /apps/ && \

--- a/pkg/app/appserver/proc.go
+++ b/pkg/app/appserver/proc.go
@@ -69,7 +69,8 @@ type Proc struct {
 
 	cmdStderr io.ReadCloser
 
-	startWg sync.WaitGroup
+	readyCh   chan struct{} // push here when ready to start app disc - protected by 'readyOnce'
+	readyOnce sync.Once     // ensures we only push to 'readyCh' once
 }
 
 // NewProc constructs `Proc`.
@@ -104,14 +105,13 @@ func NewProc(mLog *logging.MasterLogger, conf appcommon.ProcConfig, disc appdisc
 		connCh:    make(chan struct{}, 1),
 		m:         m,
 		appName:   appName,
-		startWg:   sync.WaitGroup{},
+		readyCh:   make(chan struct{}, 1),
 		cmdStderr: stderr,
 	}
 
 	if runtime.GOOS == "windows" {
 		p.ipcServerWg.Add(1)
 	}
-	p.startWg.Add(1)
 	return p
 }
 
@@ -236,7 +236,7 @@ func (p *Proc) Start() error {
 
 		go func() {
 			// App discovery start/stop.
-			p.startWg.Wait()
+			<-p.readyCh
 			p.disc.Start()
 		}()
 		defer p.disc.Stop()
@@ -334,7 +334,7 @@ func (p *Proc) SetDetailedStatus(status string) {
 	p.statusMx.Lock()
 	defer p.statusMx.Unlock()
 	if status == AppDetailedStatusRunning {
-		p.startWg.Done()
+		p.readyOnce.Do(func() { close(p.readyCh) })
 	}
 
 	if status == AppDetailedStatusRunning || status == AppDetailedStatusStopped {

